File: web-client/src/domain/auth/CurrentSession/Session/LastName.ts
import Assert from "../../../utils/Assert";
import ValidationException from "./ValidationException";

class LastName {
    public static readonly MAX_LENGTH = 50;
    public static readonly MIN_LENGTH = 1;

    public readonly value: string;

    public constructor(value: string) {
        const cleanedValue = value.trim();

        Assert.notBlank(cleanedValue, () => new LastName.MinLengthException());
        Assert.minLength(
            cleanedValue,
            LastName.MIN_LENGTH,
            () => new LastName.MinLengthException()
        );
        Assert.maxLength(
            cleanedValue,
            LastName.MAX_LENGTH,
            () => new LastName.MaxLengthException()
        );

        this.value = cleanedValue;
    }

    static readonly ValidationException = ValidationException;

    static readonly MinLengthException = class extends LastName.ValidationException {
        public constructor() {
            super(
                `Last name must be at least ${LastName.MIN_LENGTH} characters long`
            );
        }

        getCode = (): string => {
            return "MinLength";
        };
    };

    static readonly MaxLengthException = class extends LastName.ValidationException {
        public constructor() {
            super(
                `Last name must be at most ${LastName.MAX_LENGTH} characters long`
            );
        }

        getCode = (): string => {
            return "MaxLength";
        };
    };
}

export default LastName;

File: backend/server/src/main/java/com/azat4dev/demobooking/users/users_commands/domain/entities/LastName.java
package com.azat4dev.demobooking.users.users_commands.domain.entities;

import com.azat4dev.demobooking.common.DomainException;
import com.azat4dev.demobooking.common.utils.Assert;
import lombok.EqualsAndHashCode;
import lombok.Getter;

import java.io.Serializable;

@Getter
@EqualsAndHashCode
public final class LastName implements Serializable {

    public final static int MAX_LENGTH = 255;

    private final String value;

    private LastName(String value) {
        this.value = value;
    }

    private static void validate(String value) throws LastName.CantBeEmptyException, LastName.MaxLengthException {
        Assert.string(value, LastName.CantBeEmptyException::new).notNull().notBlank();
        Assert.string(value, LastName.MaxLengthException::new).maxLength(MAX_LENGTH);
    }

    public static LastName checkAndMakeFromString(String value) throws LastName.CantBeEmptyException, LastName.MaxLengthException {

        Assert.notNull(value, LastName.CantBeEmptyException::new);
        final var cleanedValue = value.trim();

        validate(cleanedValue);
        return new LastName(cleanedValue);
    }

    public static LastName dangerMakeFromStringWithoutCheck(String value) {

        assert value != null;
        return new LastName(value);
    }

    public static abstract class ValidationException extends DomainException {
        public ValidationException(String message) {
            super(message);
        }
    }

    public static class CantBeEmptyException extends ValidationException {
        public CantBeEmptyException() {
            super("Last name cannot be empty");
        }

        @Override
        public String getCode() {
            return "Empty";
        }
    }

    public static class MaxLengthException extends ValidationException {
        public MaxLengthException() {
            super("Last name is too long");
        }

        @Override
        public String getCode() {
            return "MaxLength";
        }

    }
}
